File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> um;
        for (int num: nums){
            if (um.find(num) != um.end()){
                um[num]++;
            } else {
                um[num] = 1;
            }
        }
        
        vector<pair<int, int>> data = {};
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        for (auto it=um.begin(); it != um.end(); it++){
            if (pq.size() != k){
                pq.push({it->second, it->first});
            } else {
                if (it->second > pq.top().first){
                    pq.pop();
                    pq.push({it->second, it->first});
                }
            }
        }
        vector<int> mostCommon = {};
        while (!pq.empty()){
            mostCommon.push_back(pq.top().second);
            pq.pop();
        }
        return mostCommon;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.kt
class Solution {
    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        var myMap = mutableMapOf<Int, Int>()
        for (num in nums) {
            myMap[num] = myMap.getOrDefault(num, 0) + 1
            
        }
        var pq = PriorityQueue<Pair<Int, Int>>(compareBy{ it.first })
        myMap.forEach { key, value ->
            if (pq.size < k){
                pq.add(Pair(value, key))
            } else {
                val (lastval, lastkey) = pq.poll()
                println(lastval.toString())
                println(lastkey.toString())
                println(value)
                println(key)
                if (lastval < value){
                    pq.add(Pair(value, key))
                } else {
                    println("HII")
                    pq.add(Pair(lastval, lastkey))
                }

            }
        }
        val res = IntArray(k)
        var i = 0
        while (pq.size > 0){
            res[i] = pq.poll().second
            i++
        }
        return res
    }
}